### DISCUSSION TITLE
Integration with react-dropzone

### DISCUSSION THREAD

**Discussion by eckertalex** [react-hook-form/react-hook-form#2146] [Q&A] [answered]
I need help integrating `react-hook-form` with `react-dropzone`. Here is my best try so far

```jsx
import React from 'react'
import {useForm} from 'react-hook-form'
import {DevTool} from '@hookform/devtools'
import {useDropzone} from 'react-dropzone'

export default function Form() {
  const {handleSubmit, register, control} = useForm()
  const {getRootProps, getInputProps} = useDropzone()

  function onSubmit(data) {
    console.log(data)
  }

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <div {...getRootProps()}>
        <input
          ref={register({required: true})}
          {...getInputProps({name: 'base64'})}
        />
        <p>- Drag 'n' Drop file here -</p>
      </div>
      <DevTool control={control} />
    </form>
  )
}
```

Thank you for any suggestions and great job with this library 😁

**Comment by bluebill1049** [comment, marked as answer]
try with Controller with `render` prop, all you need is to wire up the `onChange` and `value` of the component.

**Reply by eckertalex** [reply]
Thanks! That worked for me.

Her is a [CodeSandbox Example](https://codesandbox.io/s/react-hook-from-with-react-dropzone-ork6v?file=/src/App.js).
However, there is still one more problem I have with this. When I upload one file there and then try upload another, I get an error:
 
```
InvalidStateError
An attempt was made to use an object that is not, or is no longer, usable
```

**Reply by bluebill1049** [reply]
Please consider to use `render` prop for a complex component like this, then you know what you are getting onChange and what to set for value. After selecting you are getting a string of the path, i am not quite familiar with react-dropzone, you probably want to take a look at their API on what value they are expecting.

<img width="1266" alt="Screen Shot 2020-07-10 at 6 02 03 pm" src="https://user-images.githubusercontent.com/10513364/87131236-78489000-c2d7-11ea-945f-b57e4b7a8a36.png">

**Reply by eckertalex** [reply]
It turns out I did not need to provide the `input` with the `value`. I updated the CodeSandbox accordingly.

Thanks for the help!

**Reply by taschetto** [reply]
Hi, everybody!

To get the actual file instead of a string on submitting, you can do the following:

```javascript
export const DropzoneField = ({
  name,
  multiple,
  ...rest
}) => {
  const { control } = useFormContext()

  return (
    <Controller
      render={({ onChange }) => (
        <Dropzone
          multiple={multiple}
          onChange={e =>
            onChange(multiple ? e.target.files : e.target.files[0])
          }
          {...rest}
        />
      )}
      name={name}
      control={control}
      defaultValue=''
    />
  )
}

const Dropzone = ({
  multiple,
  onChange,
  ...rest
}) => {

  const {
    getRootProps,
    getInputProps,
  } = useDropzone({
    multiple,
    ...rest,
  })

  return (
    <div {...getRootProps()}>
      <input {...getInputProps({ onChange })} />
    </div>
  )
}
```

**Reply by KondapalliHandicrafts** [reply]
@taschetto  Hey I have sent the onBlur event also but the isTouched property is not set to true on blur. What i was trying to do is when he clicks on the dropzone and left i want to display error message Required.

**Reply by arellanojv** [reply]
> Thanks! That worked for me.
> 
> Her is a [CodeSandbox Example](https://codesandbox.io/s/react-hook-from-with-react-dropzone-ork6v?file=/src/App.js).
> However, there is still one more problem I have with this. When I upload one file there and then try upload another, I get an error:
> 
> ```
> InvalidStateError
> An attempt was made to use an object that is not, or is no longer, usable
> ```

@eckertalex  This code only works if you select the file manually.  react hook from doesn't detect the file when drag and drop event. Can you help to make it work for both?

**Reply by tjdam** [reply]
Is this working with drag and drop for you?
I only recently realized it works for click and search, but not drag and drop...

Is onChange supposed to be triggered when you drop a file? I don't to seem to get it to work....

**Reply by arellanojv** [reply]
No. I tried this one works for both drag n drop & click and search but it only accepts a single file.
https://stackoverflow.com/a/65809399/4184396

**Reply by doronhorwitz** [reply]
Here's a relatively strict (and [Prettier](https://prettier.io/)-ered) TypeScript version of @taschetto's [solution above](https://github.com/react-hook-form/react-hook-form/discussions/2146#discussioncomment-44167):

_(be aware that this, just as @taschetto's solution, needs to be used within a [`<FormProvider>`](https://react-hook-form.com/api/formprovider/) due to the use of [`useFormContext`](https://react-hook-form.com/api/useformcontext))._

```typescript
import { Controller, useFormContext } from "react-hook-form";
import { useDropzone } from "react-dropzone";
import { ChangeEventHandler, FC } from "react";

export const DropzoneField: FC<{ name: string; multiple?: boolean }> = ({
    name,
    multiple,
    ...rest
}) => {
    const { control } = useFormContext();

    return (
        <Controller
            render={({ field: { onChange } }) => (
                <Dropzone
                    multiple={multiple}
                    onChange={(e) =>
                        onChange(
                            multiple
                                ? e.target.files
                                : e.target.files?.[0] ?? null
                        )
                    }
                    {...rest}
                />
            )}
            name={name}
            control={control}
            defaultValue=""
        />
    );
};

const Dropzone: FC<{
    multiple?: boolean;
    onChange?: ChangeEventHandler<HTMLInputElement>;
}> = ({ multiple, onChange, ...rest }) => {
    const { getRootProps, getInputProps } = useDropzone({
        multiple,
        ...rest,
    });

    return (
        <div {...getRootProps()}>
            <input {...getInputProps({ onChange })} />
        </div>
    );
};
```

The relevant versions of packages I used (from my `package.json`):
```json
 {
    "@types/react": "^18.0.26",
    "react": "^18.2.0",
    "react-dropzone": "^14.2.3",
    "react-hook-form": "^7.41.5",
 }
```

**Reply by exipnus** [reply]
> Here's a relatively strict (and [Prettier](https://prettier.io/)-ered) TypeScript version of @taschetto's [solution above](https://github.com/react-hook-form/react-hook-form/discussions/2146#discussioncomment-44167):
> 
> _(be aware that this, just as @taschetto's solution, needs to be used within a [`<FormProvider>`](https://react-hook-form.com/api/formprovider/) due to the use of [`useFormContext`](https://react-hook-form.com/api/useformcontext))._
> 
> ```ts
> import { Controller, useFormContext } from "react-hook-form";
> import { useDropzone } from "react-dropzone";
> import { ChangeEventHandler, FC } from "react";
> 
> export const DropzoneField: FC<{ name: string; multiple?: boolean }> = ({
>     name,
>     multiple,
>     ...rest
> }) => {
>     const { control } = useFormContext();
> 
>     return (
>         <Controller
>             render={({ field: { onChange } }) => (
>                 <Dropzone
>                     multiple={multiple}
>                     onChange={(e) =>
>                         onChange(
>                             multiple
>                                 ? e.target.files
>                                 : e.target.files?.[0] ?? null
>                         )
>                     }
>                     {...rest}
>                 />
>             )}
>             name={name}
>             control={control}
>             defaultValue=""
>         />
>     );
> };
> 
> const Dropzone: FC<{
>     multiple?: boolean;
>     onChange?: ChangeEventHandler<HTMLInputElement>;
> }> = ({ multiple, onChange, ...rest }) => {
>     const { getRootProps, getInputProps } = useDropzone({
>         multiple,
>         ...rest,
>     });
> 
>     return (
>         <div {...getRootProps()}>
>             <input {...getInputProps({ onChange })} />
>         </div>
>     );
> };
> ```
> 
> The relevant versions of packages I used (from my `package.json`):
> 
> ```json
>  {
>     "@types/react": "^18.0.26",
>     "react": "^18.2.0",
>     "react-dropzone": "^14.2.3",
>     "react-hook-form": "^7.41.5",
>  }
> ```


Hello,

What is ...rest ?

Thank you

**Reply by doronhorwitz** [reply]
> 
> Hello,
> 
> What is ...rest ?
> 
> Thank you

@exipnus, I'm not sure if you're asking how `...rest` works or if you're saying that I forgot to type it.
If the former, then I suggest you go read about "Spread Syntax" and then you'll understand that using the custom `Dropzone` component above will allow you to pass down other props to react-dropzone's `useDropzone()`.
If the latter, then the typing I would use for `...rest` should be (I won't put all the code here again, just the types):

```TypeScript
export const DropzoneField: FC<{ name: string; multiple?: boolean; [x:string]: unknown;}> = ({
    name,
    multiple,
    ...rest
}) => {
    // ...
};
```
and

```TypeScript
const Dropzone: FC<{
    multiple?: boolean;
    onChange?: ChangeEventHandler<HTMLInputElement>;
    [x:string]: unknown;
}> = ({ multiple, onChange, ...rest }) => {
    // ...
};
```

**Comment by arellanojv** [comment]
Hi Bill,

I tried the solution on this discussion but this only works with the click and search option.
https://github.com/react-hook-form/react-hook-form/discussions/2146#discussioncomment-540495

**Reply by chanphiromsok** [reply]
how to get only file name? when I use react hook form with drop zone it get c:/ fakepath/"filename"

**Reply by h0t-p1xels** [reply]
I got it working properly including with both drop and click to add a file  using the following code:

**FileInput.js**
```
import React, { useCallback, useEffect } from "react"
import { useDropzone } from "react-dropzone"
import { useFormContext } from "react-hook-form"

const FileInput = props => {
    const { name, label = name } = props
    const { register, unregister, setValue, watch } = useFormContext()
    const files = watch(name)
    const onDrop = useCallback(
        droppedFiles => {
            setValue(name, droppedFiles, { shouldValidate: true })
        },
        [setValue, name]
    )
    const { getRootProps, getInputProps, isDragActive } = useDropzone({
        onDrop,
        accept: props.accept,
    })
    useEffect(() => {
        register(name)
        return () => {
            unregister(name)
        }
    }, [register, unregister, name])
    return (
        <>
            <label className=" " htmlFor={name}>
                {label}
            </label>
            <div
                {...getRootProps()}
                type="file"
                role="button"
                aria-label="File Upload"
                id={name}
            >
                <input {...props} {...getInputProps()} />
                <div
                    style={{ width: "500px", border: "black solid 2px" }}
                    className={" " + (isDragActive ? " " : " ")}
                >
                    <p className=" ">Drop the files here ...</p>

                    {!!files?.length && (
                        <div className=" ">
                            {files.map(file => {
                                return (
                                    <div key={file.name}>
                                        <img
                                            src={URL.createObjectURL(file)}
                                            alt={file.name}
                                            style={{
                                                height: "200px",
                                            }}
                                        />
                                    </div>
                                )
                            })}
                        </div>
                    )}
                </div>
            </div>
        </>
    )
}

export default FileInput
```


**Form.js**
```
import React from "react"
import { FormProvider, useForm } from "react-hook-form"
import FileInput from "./FileInput"

const Form = () => {
    const methods = useForm({
        mode: "onBlur",
    })
    const onSubmit = methods.handleSubmit(values => {
        console.log("values", values)
    })

    return (
        <FormProvider {...methods}>
            <form onSubmit={onSubmit}>
                <div className="">
                    <FileInput
                        accept="image/png, image/jpg, image/jpeg, image/gif"
                        name="file alt text"
                        label="File Upload"
                    />
                </div>
            </form>
        </FormProvider>
    )
}

export default Form
```

**Reply by chanphiromsok** [reply]
thank you

**Reply by bluebill1049** [reply]
Thank you for sharing solutions.

**Reply by elebumm** [reply]
@nikkipantony You saved me hours and hours of time. Thank you so much for your solution.

**Reply by Tan-Aki** [reply]
@nikkipantony  You are a lifesaver.

As for those who wonder how to be able to add multiples files, one after the other (and not replacing the uploaded files each time).

Replace the onDrop function with :

```
    const onDrop = useCallback(
        (droppedFiles) => {
            const newFiles = (!!files?.length && [...files].concat(droppedFiles)) || droppedFiles;
            setValue(name, newFiles, { shouldValidate: true });
        },
        [setValue, name, files],
    );
```

**Reply by nemanjam** [reply]
> I got it working properly including with both drop and click to add a file using the following code:
> 
> **FileInput.js**
> 
> ```
> import React, { useCallback, useEffect } from "react"
> import { useDropzone } from "react-dropzone"
> import { useFormContext } from "react-hook-form"
> 
> const FileInput = props => {
>     const { name, label = name } = props
>     const { register, unregister, setValue, watch } = useFormContext()
>     const files = watch(name)
>     const onDrop = useCallback(
>         droppedFiles => {
>             setValue(name, droppedFiles, { shouldValidate: true })
>         },
>         [setValue, name]
>     )
>     const { getRootProps, getInputProps, isDragActive } = useDropzone({
>         onDrop,
>         accept: props.accept,
>     })
>     useEffect(() => {
>         register(name)
>         return () => {
>             unregister(name)
>         }
>     }, [register, unregister, name])
>     return (
>         <>
>             <label className=" " htmlFor={name}>
>                 {label}
>             </label>
>             <div
>                 {...getRootProps()}
>                 type="file"
>                 role="button"
>                 aria-label="File Upload"
>                 id={name}
>             >
>                 <input {...props} {...getInputProps()} />
>                 <div
>                     style={{ width: "500px", border: "black solid 2px" }}
>                     className={" " + (isDragActive ? " " : " ")}
>                 >
>                     <p className=" ">Drop the files here ...</p>
> 
>                     {!!files?.length && (
>                         <div className=" ">
>                             {files.map(file => {
>                                 return (
>                                     <div key={file.name}>
>                                         <img
>                                             src={URL.createObjectURL(file)}
>                                             alt={file.name}
>                                             style={{
>                                                 height: "200px",
>                                             }}
>                                         />
>                                     </div>
>                                 )
>                             })}
>                         </div>
>                     )}
>                 </div>
>             </div>
>         </>
>     )
> }
> 
> export default FileInput
> ```
> 
> **Form.js**
> 
> ```
> import React from "react"
> import { FormProvider, useForm } from "react-hook-form"
> import FileInput from "./FileInput"
> 
> const Form = () => {
>     const methods = useForm({
>         mode: "onBlur",
>     })
>     const onSubmit = methods.handleSubmit(values => {
>         console.log("values", values)
>     })
> 
>     return (
>         <FormProvider {...methods}>
>             <form onSubmit={onSubmit}>
>                 <div className="">
>                     <FileInput
>                         accept="image/png, image/jpg, image/jpeg, image/gif"
>                         name="file alt text"
>                         label="File Upload"
>                     />
>                 </div>
>             </form>
>         </FormProvider>
>     )
> }
> 
> export default Form
> ```

This stopped to work with React 18, I am struggling with it now. It worked perfectly with React 17.

**Reply by Moshyfawn** [reply]
RHF V7 does not officially support React 18. You can try to install [the next major V8 alpha version](https://github.com/react-hook-form/react-hook-form/releases/tag/v8.0.0-alpha.2) and see if it fixes the issue.

Also, consider the fact that the bottleneck can also be [the react-dropzone package](https://github.com/react-dropzone/react-dropzone/issues). I saw quite a few React 18 related issue there.

React 18 just came out, so give the library maintainers some time to adjust to this breaking major React update. It's gonna a while before libraries can work with React 18 bug-free

**Reply by nemanjam** [reply]
I didnt know that. What about this?

And for React Query v4 I need React 18. What a mess???

Edit: I tried `8.0.0-alpha.2`, it doesn't work, breaks same way, image file is lost from form state on subsequent renders.

```ts
  useEffect(() => {
    register(name);

    return () => {
      unregister(name); // this line breaks image files to load in form state.
    };
  }, [register, unregister, name]);
```

![image](https://user-images.githubusercontent.com/9990165/162636932-c7d5b401-5afc-47fa-8e33-6397bd58d67f.png)

**Reply by nemanjam** [reply]
Actually it load files but doesn't keep formState properly, dirty is lost for example.

![image](https://user-images.githubusercontent.com/9990165/162676156-812bc35e-c938-485a-af08-3c73d1a08a15.png)

**Comment by michaeloliverx** [comment]
Here is TypeScript example using the `<Controller>` component:
```tsx
import { useForm, SubmitHandler, Controller } from 'react-hook-form';
import Dropzone from 'react-dropzone';

type FormInputs = {
  file: FileList;
};

export default function App() {
  const form = useForm<FormInputs>();

  const onSubmit: SubmitHandler<FormInputs> = (data) => {
    console.log(data);
  };

  return (
    <div>
      <form onSubmit={form.handleSubmit(onSubmit)}>
        <Controller
          control={form.control}
          name="file"
          rules={{
            required: { value: true, message: 'This field is required' },
          }}
          render={({ field: { onChange, onBlur }, fieldState }) => (
            <Dropzone
              noClick
              onDrop={(acceptedFiles) => {
                form.setValue('file', acceptedFiles as unknown as FileList, {
                  shouldValidate: true,
                });
              }}
            >
              {({
                getRootProps,
                getInputProps,
                open,
                isDragActive,
                acceptedFiles,
              }) => (
                <div>
                  <div
                    style={{
                      borderStyle: 'dashed',
                      backgroundColor: isDragActive ? `#808080` : 'transparent',
                    }}
                    {...getRootProps()}
                  >
                    <input
                      {...getInputProps({
                        id: 'spreadsheet',
                        onChange,
                        onBlur,
                      })}
                    />

                    <p>
                      <button type="button" onClick={open}>
                        Choose a file
                      </button>{' '}
                      or drag and drop
                    </p>

                    {acceptedFiles.length
                      ? acceptedFiles[0].name
                      : 'No file selected.'}

                    <div>
                      {fieldState.error && (
                        <span role="alert">{fieldState.error.message}</span>
                      )}
                    </div>
                  </div>
                </div>
              )}
            </Dropzone>
          )}
        />

        <button>Submit</button>
      </form>
    </div>
  );
}

```

**Reply by AhsanTariq97** [reply]
How do you make it accept only specific file types?

**Reply by noxify** [reply]
@AhsanTariq97 - Just add `accept` to the `Dropzone` element.

```ts
<Dropzone 
   noClick
  accept={{"text/csv":[]}}
  onDrop={(acceptedFiles) => {
    form.setValue('file', acceptedFiles as unknown as FileList, {
      shouldValidate: true,
    });
  }}
>
...
```

Docs: https://react-dropzone.js.org/#src

**Reply by AhsanTariq97** [reply]
I found a workaround for it in the meanwhile, but this is the right way to do it. Thanks

**Reply by mengxi-ream** [reply]
If you use onDrop, then you don't need Controller, see [this example](https://codesandbox.io/p/sandbox/react-hook-form-uploadfile-with-thumbnail-preview--drag-and-drop-p11gy?file=%2Fsrc%2FFileInput.jsx%3A34%2C30-34%2C56)

**Reply by jgustavoj** [reply]
Anyone have this example with appended files?

**Comment by lamhungypl** [comment]
I notice that these working samples are relied on `onDrop` props. 
ReactDropzone doesn't expose files values on their input's  `ref,  onChange`, isn't it?

**Comment by jackhkmatthewsBB** [comment]
I was experiencing some related unexpected behaviour.

The `onChange` handler is not called for inputs wrapped with `getRootProps()`.

Instead pass an `onDrop` callback to `useDropzone`.

This works:

```tsx
function Form() {
  const { register, handleSubmit, setValue } = useForm({});

  const dropzone = useDropzone({
    onDrop: ([file]) => file && setValue("file", file),
  });

  return (
    <form onSubmit={handleSubmit(console.log)}>
      <div {...dropzone.getRootProps()}>
        Drop here
        <input
          {...dropzone.getInputProps({
            ...register("file"),
          })}
        />
      </div>
      <button type="submit">Upload</button>
    </form>
  );
};
```

This doesn't:

```tsx
function Form() {
  const { register, control, handleSubmit, setValue } = useForm({});

  const dropzone = useDropzone();

  return (
    <form onSubmit={handleSubmit(console.log)}>
      <Controller
        control={control}
        name="file"
        render={({ field: { onBlur } }) => {
          return (
            <div {...dropzone.getRootProps()}>
              Drop here
              <input
                {...dropzone.getInputProps({
                  onChange: (e) => setValue(e.target.files),
                  onBlur,
                })}
              />
            </div>
          );
        }}
      />
      <button type="submit">Upload</button>
    </form>
  );
}

```

**Reply by lagroms** [reply]
I have the same problem, I had to change my code with your exemple using "register". Is that some sort of bug that need to be opened? I upgraded recently from v12 to v14, and I didn't have this issue before.

**Comment by mengxi-ream** [comment]
I found a really great solution that uses `onDrop` and doesn't require `onChange` or anything like that. Here is the [code](https://codesandbox.io/p/sandbox/react-hook-form-uploadfile-with-thumbnail-preview--drag-and-drop-p11gy?file=%2Fsrc%2FFileInput.jsx%3A34%2C30-34%2C56). I believe this should be chosen as the answer!

**Reply by mengxi-ream** [reply]
Here is the core code, it even supports append files!😃
```jsx
import React, { useCallback, useEffect } from "react";
import { useDropzone } from "react-dropzone";
import { useForm } from "react-hook-form";

const FileInput = (props) => {
  const { name, label = name, mode = "update" } = props;

  const { register, unregister, setValue, watch } = useForm();

  const files = watch(name);

  const onDrop = useCallback(
    (droppedFiles) => {
      let newFiles =
        mode === "update" ? droppedFiles : [...(files || []), ...droppedFiles];
      if (mode === "append") {
        newFiles = newFiles.reduce((prev, file) => {
          const fo = Object.entries(file);
          if (
            prev.find((e) => {
              const eo = Object.entries(e);
              return eo.every(
                ([key, value], index) =>
                  key === fo[index][0] && value === fo[index][1]
              );
            })
          ) {
            return prev;
          } else {
            return [...prev, file];
          }
        }, []);
      }
      setValue(name, newFiles, { shouldValidate: true });
    },
    [setValue, name, mode, files]
  );

  const { getRootProps, getInputProps, isDragActive } = useDropzone({
    onDrop,
    accept: props.accept
  });

  useEffect(() => {
    register(name);
    return () => {
      unregister(name);
    };
  }, [register, unregister, name]);

  return (
    <div style={{ padding: "100px" }}>
      <label
        className="block text-gray-700 text-sm font-bold mb-2 capitalize"
        htmlFor={name}
      >
        {label}
      </label>

      <div {...getRootProps()}>
        <input
          {...props}
          className="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
          id={name}
          {...getInputProps()}
        />

        <div
          className={
            "w-full p-2 border border-dashed border-gray-900 " +
            (isDragActive ? "bg-gray-400" : "bg-gray-200")
          }
        >
          <p className="text-center my-2">Drop the files here ...</p>
          {!!files?.length && (
            <div className="grid gap-1 grid-cols-4 mt-2">
              {files.map((file) => {
                return (
                  <div key={file.name}>
                    <img
                      src={URL.createObjectURL(file)}
                      alt={file.name}
                      style={{ width: "100px", height: "100px" }}
                    />
                  </div>
                );
              })}
            </div>
          )}
        </div>
      </div>
    </div>
  );
};

export default FileInput;

```

**Comment by epsilon11101** [comment]
With Mui and React Hook Form, this can work, adapt it only to your needs, I omitted some code that I can't publish, I hope it helps someone.


```jsx
import {
  Controller,
  FieldValues,
  FieldPath,
  UseFormReturn,
  RegisterOptions,
  FieldErrors,
  FieldName,
} from 'react-hook-form'

import {
  ErrorMessage,
  FieldValuesFromFieldErrors,
} from '@hookform/error-message'



import {
  Box,
  Button,
  Card,
  CardContent,
  Grid,
  LinearProgress,
  ListItem,
  ListItemIcon,
  ListItemText,
  Stack,
  Typography,
  useTheme,
} from '@mui/material'
import Dropzone, { DropzoneProps } from 'react-dropzone'
import FileUploadIcon from '@mui/icons-material/FileUpload'
import DescriptionIcon from '@mui/icons-material/Description'
import { useCallback, useState } from 'react'

interface Props<TFieldValues extends FieldValues>
  extends Omit<DropzoneProps, 'onDrop'> {
  name: FieldPath<TFieldValues>
  control: UseFormReturn<TFieldValues>['control']
  rules?: RegisterOptions
  onChange?: (value: string | null) => void
  errors: FieldErrors<TFieldValues>
}

const FormDrag = <TFieldValues extends FieldValues>({
  name,
  control,
  rules,
  onChange,
  errors,
  ...rest
}: Props<TFieldValues>) => {

  const theme = useTheme()
  const [file, setFile] = useState<File | null>(null)
  const [isUploading, setIsUploading] = useState(false)
  const [finished, setFinished] = useState(false)

  const onReset = () => {
    setIsUploading(false)
    setFinished(false)
    setFile(null)
  }

  const onDropHandler = useCallback(
    (acceptedFiles: File[], onControllerChange: (value: string) => void) => {
      const uploadFile = Array.isArray(acceptedFiles)
        ? acceptedFiles[0]
        : acceptedFiles
      if (uploadFile) {
        setFinished(false)
        setFile(uploadFile)
        setIsUploading(true)
        fetchSomething({....})
          .unwrap()
          .then(({ .......}) => {
            otherPrivateFetch({ ...... })
              .unwrap()
              .then(() => {
                onControllerChange("whateveryouneed")
                onChange && onChange("whateveryouneed")
                setIsUploading(false)
                setFinished(true)
              })
              .catch(() => onReset())
          })
          .catch(() => onReset())
      }
    },
    []
  )

  return (
    <Controller
      name={name}
      control={control}
      rules={rules}
      render={({ field: { onChange: onControllerChange, onBlur } }) => (
        <Grid item container>
          {!isUploading && file === null && (
            <Dropzone
              onDrop={(acceptedFiles) => {
                onDropHandler(acceptedFiles, onControllerChange)
              }}
              {...rest}
            >
              {({ getRootProps, getInputProps, open, isDragActive }) => {
                return (
                  <Card
                    variant="outlined"
                    component="div"
                    elevation={0}
                    sx={{
                      px: 4,
                      width: '294px',
                      height: '195px',
                      backgroundColor: theme.palette.gray.main,
                    }}
                    {...getRootProps()}
                  >
                    <CardContent>
                      <Box
                        component="input"
                        {...getInputProps({
                          onBlur,
                        })}
                      />
                      <Grid
                        item
                        container
                        direction="column"
                        justifyContent="center"
                        alignItems="strech"
                        rowGap={2}
                        wrap="nowrap"
                      >
                        <Grid item xs={12}>
                          <Stack
                            direction="column"
                            width="100%"
                            justifyContent="center"
                            alignItems="center"
                            gap={1}
                          >
                            <FileUploadIcon color="darkGrey" />
                            <Typography
                              variant="body1"
                              component="p"
                              textTransform={
                                !isDragActive ? 'capitalize' : 'uppercase'
                              }
                              textAlign="center"
                              color={theme.palette.darkGrey.main}
                            >
                              {!isDragActive
                                ? 'firstText'
                                : 'Dragtext'}
                            </Typography>
{.......morecomponents}
                          </Stack>
                        </Grid>
                        <Grid
                          item
                          container
                          justifyContent="center"
                          alignItems="strech"
                          xs={12}
                          sx={{
                            display: !isDragActive ? 'block' : 'none',
                          }}
                        >
                          <Button
                            fullWidth
                            type="button"
                            onClick={open}
                            variant="outlined"
                            sx={{
                              py: 1,
                              px: 2,
                              color: 'common.black',
                              border: `3px solid ${theme.palette.primary.main}`,
                              borderRadius: '15px',
                              '&:hover': {
                                borderWidth: '3px',
                                backgroundColor: theme.palette.primary.main,
                                color: theme.palette.primary.contrastText,
                              },
                            }}
                          >
                           {..........button_text}
                          </Button>
                        </Grid>
                      </Grid>
                    </CardContent>
                  </Card>
                )
              }}
            </Dropzone>
          )}
          {!!file && (
            <Grid
              item
              container
              xs={12}
              direction="row"
              justifyContent="flex-start"
              alignItems="center"
              sx={{ pl: 2 }}
            >
              <Grid item xs={12}>
                <ListItem disablePadding disableGutters>
                  <ListItemIcon sx={{ minWidth: 0, mr: 2 }}>
                    <DescriptionIcon />
                  </ListItemIcon>
                  <ListItemText
                    primary={file.name}
                    primaryTypographyProps={{
                      variant: 'subtitle1',
                      fontWeight: 600,
                    }}
                  />
                </ListItem>

                <LinearProgress
                  sx={{
                    marginTop: 0.5,
                    maxWidth: '294px',
                  }}
                  variant="determinate"
                  value={finished && !isUploading ? 100 : 0}
                />
              </Grid>
            </Grid>
          )}
          {errors && (
            <Grid item xs={12} sx={{ mt: 1 }}>
              <ErrorMessage
                errors={errors}
                name={
                  name as unknown as FieldName<
                    FieldValuesFromFieldErrors<FieldErrors<TFieldValues>>
                  >
                }
                render={({ message }) => {
                  return (
                    <Typography
                      variant="body2"
                      component="p"
                      color={theme.palette.error.main}
                    >
                      {message}
                    </Typography>
                  )
                }}
              />
            </Grid>
          )}
        </Grid>
      )}
    />
  )
}

export default FormDrag
```

in Client code you can use like this 

 ```jsx
     <FormDrag
        name="passwordKey"
        control={control}
        errors={errors}
        documentId={3}
        noClick
         noKeyboard
         maxFiles={1}/>
```
